### ISSUE TITLE
Expanding the "Recent commits" section sometimes doesn't work

### ISSUE THREAD

**Issue by AThePeanut4** [NeogitOrg/neogit#1150]
### Description

As per title - I've been running into this issue for ages and I've finally managed to figure out a consistent way to cause the issue.

### Neovim version

NVIM v0.10.0-dev-2191+g4e59422e1d
Build type: RelWithDebInfo
LuaJIT 2.1.1702233742

### Operating system and version

Arch Linux

### Steps to reproduce

1. Open nvim in a git repo with a few commits, and one stash
2. Open Neogit
3. Expand, and then collapse, the "Recent commits" section
4. Expand the "Stashes" section
5. Try to expand the "Recent commits" section, and observe that nothing happens

This also happens in other situations, after expanding unstaged/changes as well - these are just the simplest steps to consistently trigger the bug.

### Expected behavior

Expanding the "Recent commits" section should work.

### Actual behavior

Here's a screen recording demonstrating the issue:

https://github.com/NeogitOrg/neogit/assets/49614525/bfc40e46-1425-489e-abb9-bbd2281dd624

### Minimal config

```Lua
-- No changes made

-- NOTE: See the end of this file if you are reporting an issue, etc. Ignore all the "scary" functions up top, those are
-- used for setup and other operations.
local M = {}

local base_root_path = vim.fn.fnamemodify(debug.getinfo(1, "S").source:sub(2), ":p:h") .. "/.min"
function M.root(path)
  return base_root_path .. "/" .. (path or "")
end

function M.load_plugin(plugin_name, plugin_url)
  local package_root = M.root("plugins/")
  local install_destination = package_root .. plugin_name
  vim.opt.runtimepath:append(install_destination)

  if not vim.loop.fs_stat(package_root) then
    vim.fn.mkdir(package_root, "p")
  end

  if not vim.loop.fs_stat(install_destination) then
    print(string.format("> Downloading plugin '%s' to '%s'", plugin_name, install_destination))
    vim.fn.system({
      "git",
      "clone",
      "--depth=1",
      plugin_url,
      install_destination,
    })
    if vim.v.shell_error > 0 then
      error(string.format("> Failed to clone plugin: '%s' in '%s'!", plugin_name, install_destination),
        vim.log.levels.ERROR)
    end
  end
end

---@alias PluginName string The plugin name, will be used as part of the git clone destination
---@alias PluginUrl string The git url at which a plugin is located, can be a path. See https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols for details
---@alias MinPlugins table<PluginName, PluginUrl>

---Do the initial setup. Downloads plugins, ensures the minimal init does not pollute the filesystem by keeping
---everything self contained to the CWD of the minimal init file. Run prior to running tests, reproducing issues, etc.
---@param plugins? table<PluginName, PluginUrl>
function M.setup(plugins)
  vim.opt.packpath = {}                      -- Empty the package path so we use only the plugins specified
  vim.opt.runtimepath:append(M.root(".min")) -- Ensure the runtime detects the root min dir

  -- Install required plugins
  if plugins ~= nil then
    for plugin_name, plugin_url in pairs(plugins) do
      M.load_plugin(plugin_name, plugin_url)
    end
  end

  vim.env.XDG_CONFIG_HOME = M.root("xdg/config")
  vim.env.XDG_DATA_HOME = M.root("xdg/data")
  vim.env.XDG_STATE_HOME = M.root("xdg/state")
  vim.env.XDG_CACHE_HOME = M.root("xdg/cache")

  -- NOTE: Cleanup the xdg cache on exit so new runs of the minimal init doesn't share any previous state, e.g. shada
  vim.api.nvim_create_autocmd("VimLeave", {
    callback = function()
      vim.fn.system({
        "rm",
        "-r",
        "-f",
        M.root("xdg")
      })
    end
  })
end

-- NOTE: If you have additional plugins you need to install to reproduce your issue, include them in the plugins
-- table within the setup call below.
M.setup({
  plenary = "https://github.com/nvim-lua/plenary.nvim.git",
  telescope = "https://github.com/nvim-telescope/telescope.nvim",
  diffview = "https://github.com/sindrets/diffview.nvim",
  neogit = "https://github.com/NeogitOrg/neogit"
})
-- WARN: Do all plugin setup, test runs, reproductions, etc. AFTER calling setup with a list of plugins!
-- Basically, do all that stuff AFTER this line.
require("neogit").setup({}) -- For instance, setup Neogit
```

**Comment by igorlfs** [comment]
Also having this issue on the nightly branch

**Comment by CKolkey** [comment]
Are you sure? I'm not able to reproduce it.

**Comment by AThePeanut4** [comment]
@CKolkey The original steps seem to no longer be able to consistently reproduce the issue, but I can confirm it is still happening. Here are new steps:

1. Open nvim in a clean clone of the neogit repo.
2. We need an unstaged change, I'm not sure if the length or number of changes matters. What I did was use a formatter to remove all trailing whitespace from `README.md` :upside_down_face:. Doing `:%s/\s\+$//` also works.
3. Save the changes (`:w`)
4. Open Neogit
5. Expand, and then collapse, the "Recent commits" section
6. Expand the diff for README.md under the Unstaged changes section
7. Try to expand and then collapse the "Recent commits" section. There's something very weird going on here - it sometimes works and sometimes doesn't, depending on if I scroll with the mouse, click with the mouse on the heading, use `<C-D>`, or move just the cursor with `j`.

When #1161 merged I quickly tried it out, but the status buffer was super broken (iirc, it would crash while trying to close, among other issues), so I just went back to using my local fork (#1151 merged into nightly). Unfortunately I didn't have the time to investigate all the issues and make bug reports for them.

After quickly testing it just now, it seems like most of the bigger issues have been fixed, but some still remain. Hopefully I'll get around to making a bug report some time this week.

**Comment by CKolkey** [comment]
Great reproduction - I'll have a peek now. And if you have other issues with the new status buffer, I'd love bug reports :D

**Comment by CKolkey** [comment]
Easy fix for the issue as you describe: [`10d1fe0` (#978)](https://github.com/NeogitOrg/neogit/pull/978/commits/10d1fe02a8d183fda2cc1997cda28bfef3ebc10b)

The issue is that, depending on how you scroll down, your cursor may be on the last (blank) line of the folded section, though it will appear as though it's on the fold. This was caught by an early return that serves to prevent the blank line between sections from being togglable.

**Comment by igorlfs** [comment]
Hey fellas, couldn't catch up earlier! I haven't tested extensively, but with the latest commit on nightly, I'm no longer having this issue

**Comment by AThePeanut4** [comment]
@CKolkey Thanks for the fix! I'm still getting some very occasional weirdness when expanding/collapsing folds (all folds, not just the "Recent commits" section), but only after having used Neogit for a while. I'll probably have to do some code digging to work out reproduction steps, like I had to do with the original issue here.

**Comment by CKolkey** [comment]
Cool, I'll close this then, but any/all bug reports are welcome. I use this daily at work, so any issues I find, I fix, but that's just _my_ workflow.
